Sources/NIOSSL/SSLContext.swift

```diff
@@ -112,6 +112,101 @@ private func alpnCallback(ssl: OpaquePointer?,
+          let outputPSK = psk // Output PSK key.
+          else {
+        return 0
+    }
+    
+    // Take the hint and the possible identity and pass them down to the callback to get associated PSK from callback
+    guard let contextPSKIdentityCallback = try? serverCallback(configurationHint, strIdentity) else {
+        return 0
+    }
+    let serverPSK = contextPSKIdentityCallback.key // From the callback
+    
+    // Make sure the key is returned by the callback and it is of proper length, otherwise, fail.
+    if serverPSK.isEmpty || serverPSK.count > max_psk_len {
+        return 0
+    }
+    let _ = serverPSK.withUnsafeBytes { (body: UnsafeRawBufferPointer) -> Void in
+        guard let unsafeRawPointer = body.baseAddress else { return }
+        memcpy(outputPSK, unsafeRawPointer, body.count)
+    }
+    return UInt32(serverPSK.count)
+}
+
+/// PSK Callback for the client side context.
+private func clientPSKCallback(ssl: OpaquePointer?,
+                                 hint: UnsafePointer<CChar>?,
+                                 identity: UnsafeMutablePointer<CChar>?,
+                                 max_identity_len: UInt32,
+                                 psk: UnsafeMutablePointer<UInt8>?,
+                                 max_psk_len: UInt32) -> UInt32 {
+    
+    guard let ssl = ssl else { return 0 }
+    // Initial implementation only supports TLS 1.2 due API support exposed from BoringSSL.
+    // TODO (meaton) add TLS 1.3 support when available.
+    
+    // We want to take the SSL pointer and extract the parent Swift object.
+    let parentCtx = CNIOBoringSSL_SSL_get_SSL_CTX(ssl)!
+    let parentPtr = CNIOBoringSSLShims_SSL_CTX_get_app_data(parentCtx)!
+    let parentSwiftContext: NIOSSLContext = Unmanaged.fromOpaque(parentPtr).takeUnretainedValue()
+    
+    guard let clientCallback = parentSwiftContext.pskClientConfigurationCallback,
+          let unwrappedIdentity = identity, // Output identity that will be later be mapped from client callback
+          let outputPSK = psk // Output PSK key that will later be mapped from client callback
+          else {
+        return 0
+    }
+    
+    // If set, build out a hint to pass into the client callback.
+    guard let clientHint = hint,
+          let derivedHint = String(validatingUTF8: clientHint) else {
+        return 0
+    }
+    // Take the hint and the possible identity and pass them down to the callback to get associated PSK from callback
+    guard let pskIdentityCallback = try? clientCallback(derivedHint) else {
+        return 0
+    }
+    let clientPSK = pskIdentityCallback.key // Key from the callback
+    let clientIdentity = pskIdentityCallback.identity ?? ""
+    
+    if clientIdentity.isEmpty {
+        return 0
+    }
+    // Map the output identity from the one passed back from the callback.
+    // This helps populate the server callback for the key exchange.
+    memcpy(unwrappedIdentity, clientIdentity, clientIdentity.count)
```

Comment:
This algorithm isn't right, we need to use `withCString` here. This has two issues: first, it uses `clientIdentity.count`, which is a count of `String.Character`, not of bytes; and second, we are required to write a null-terminator into this buffer per the doc-string.

Relatedly, we also aren't comparing against `max_identity_len`, so this can trigger a buffer overrun.

Reply:
Addressed in 0101dcb, let me know if this is what you had in mind.